### ISSUE TITLE
Support loaders

### ISSUE THREAD

**Issue by RyanTheAllmighty** [ATLauncher/ATLauncher#338]
Currently with Forge for 1.13.2 and later, Forge no longer provides a simple way to install the libraries it needs.

It needs to run a bunch of processes on the users machine to generate mappings, patches etc.

In order to support this, and going forward more loaders, including Fabric, I want to remove the forge libraries being defined in the pack xml definitions and switch to the below:

```xml
<loader type="forge" version="25.0.191" />
```

The launcher will see that and then download and install forge libraries and run the processes it needs to install for a pack on the launcher client, rather than pack developers needing to worry about the libraries themselves.

This will require a bit of work as the launcher doesn't currently do anything specific for Forge or any other loaders.

The process is as follows:

- Download https://files.minecraftforge.net/maven/net/minecraftforge/forge/1.13.2-25.0.191/forge-1.13.2-25.0.191-installer.jar
- Extract the file `install_profile.json` and `version.json` out of the jar
- Download the libraries from the `libraries` section of `install_profile.json` (support for this is the same as Vanilla Minecraft so should be little effort). These libraries are for the install of Forge.
- Run the processors from the `processors` part of `install_profile.json` to produce the outputs
- Download the libraries from the `libraries` section of `version.json` (support for this is the same as Vanilla Minecraft so should be little effort). These libraries are for the game to run
- The forge installer jar also contains `maven/net/minecraftforge/forge/{mcversion}-{forgeversion}/` directory which contains the forge jar and universal jar.

While it seems like a lot of work, it really shouldn't be as the work I've done in the develop branch on the weekend supports the downloading of libraries already, just need to get the processors running.

**Comment by RyanTheAllmighty** [comment]
Example `install_profile.json`:

```json
{
    "_comment_": [
        "Please do not automate the download and installation of Forge.",
        "Our efforts are supported by ads from the download page.",
        "If you MUST automate this, please consider supporting the project through https://www.patreon.com/LexManos/"
    ],
    "spec": 0,
    "profile": "forge",
    "version": "1.13.2-forge-25.0.191",
    "json": "/version.json",
    "path": "net.minecraftforge:forge:1.13.2-25.0.191",
    "logo": "/big_logo.png",
    "minecraft": "1.13.2",
    "welcome": "Welcome to the simple Forge installer.",
    "data": {
        "MAPPINGS": {
            "client": "[de.oceanlabs.mcp:mcp_config:1.13.2-20190213.203750:mappings@txt]",
            "server": "[de.oceanlabs.mcp:mcp_config:1.13.2-20190213.203750:mappings@txt]"
        },
        "BINPATCH": {
            "client": "/data/client.lzma",
            "server": "/data/server.lzma"
        },
        "MC_SLIM": {
            "client": "[net.minecraft:client:1.13.2:slim]",
            "server": "[net.minecraft:server:1.13.2:slim]"
        },
        "MC_SLIM_SHA": {
            "client": "'46c94ec3485ec35fe1826ca17ecfd5e10bb17c9c'",
            "server": "'327e89f60e69407b07c99887872bc05ed8f3b865'"
        },
        "MC_EXTRA": {
            "client": "[net.minecraft:client:1.13.2:extra]",
            "server": "[net.minecraft:server:1.13.2:extra]"
        },
        "MC_EXTRA_SHA": {
            "client": "'938487f223e08b48c7b40dc7aa40a34b9dea2520'",
            "server": "'d79d1c88b752a7ed472b20317b123acc720a4089'"
        },
        "MC_SRG": {
            "client": "[net.minecraft:client:1.13.2-20190213.203750:srg]",
            "server": "[net.minecraft:server:1.13.2-20190213.203750:srg]"
        },
        "PATCHED": {
            "client": "[net.minecraftforge:forge:1.13.2-25.0.191:client]",
            "server": "[net.minecraftforge:forge:1.13.2-25.0.191:server]"
        },
        "PATCHED_SHA": {
            "client": "'c513c6e183d3a6d4fd9bbadca5a9408f4aa04abe'",
            "server": "'1a6d2aa4f052990d144386c1abff849b787fa408'"
        }
    },
    "processors": [
        {
            "jar": "net.minecraftforge:installertools:1.0.3",
            "classpath": [
                "net.sf.jopt-simple:jopt-simple:5.0.4",
                "com.google.code.gson:gson:2.8.0"
            ],
            "args": [
                "--task",
                "MCP_DATA",
                "--input",
                "[de.oceanlabs.mcp:mcp_config:1.13.2-20190213.203750@zip]",
                "--output",
                "{MAPPINGS}",
                "--key",
                "mappings"
            ]
        },
        {
            "jar": "net.minecraftforge:jarsplitter:1.1.0",
            "classpath": [
                "net.sf.jopt-simple:jopt-simple:5.0.4"
            ],
            "args": [
                "--input",
                "{MINECRAFT_JAR}",
                "--slim",
                "{MC_SLIM}",
                "--extra",
                "{MC_EXTRA}",
                "--srg",
                "{MAPPINGS}"
            ],
            "outputs": {
                "{MC_SLIM}": "{MC_SLIM_SHA}",
                "{MC_EXTRA}": "{MC_EXTRA_SHA}"
            }
        },
        {
            "jar": "net.minecraftforge:installertools:1.0.3",
            "classpath": [
                "net.sf.jopt-simple:jopt-simple:5.0.4",
                "com.google.code.gson:gson:2.8.0"
            ],
            "args": [
                "--task",
                "CREATE_PARENTS",
                "--target",
                "{MC_SRG}"
            ]
        },
        {
            "jar": "net.md-5:SpecialSource:1.8.3",
            "classpath": [
                "org.ow2.asm:asm-commons:6.1.1",
                "net.sf.jopt-simple:jopt-simple:4.9",
                "com.google.guava:guava:20.0",
                "net.sf.opencsv:opencsv:2.3",
                "org.ow2.asm:asm-analysis:6.1.1",
                "org.ow2.asm:asm-tree:6.1.1",
                "org.ow2.asm:asm:6.1.1"
            ],
            "args": [
                "--in-jar",
                "{MC_SLIM}",
                "--out-jar",
                "{MC_SRG}",
                "--srg-in",
                "{MAPPINGS}"
            ]
        },
        {
            "jar": "net.minecraftforge:binarypatcher:1.0.5",
            "classpath": [
                "commons-io:commons-io:2.4",
                "com.google.guava:guava:25.1-jre",
                "net.sf.jopt-simple:jopt-simple:5.0.4",
                "com.github.jponge:lzma-java:1.3",
                "com.nothome:javaxdelta:2.0.1",
                "com.google.code.findbugs:jsr305:3.0.2",
                "org.checkerframework:checker-qual:2.0.0",
                "com.google.errorprone:error_prone_annotations:2.1.3",
                "com.google.j2objc:j2objc-annotations:1.1",
                "org.codehaus.mojo:animal-sniffer-annotations:1.14",
                "trove:trove:1.0.2"
            ],
            "args": [
                "--clean",
                "{MC_SRG}",
                "--output",
                "{PATCHED}",
                "--apply",
                "{BINPATCH}"
            ],
            "outputs": {
                "{PATCHED}": "{PATCHED_SHA}"
            }
        }
    ],
    "libraries": [
        {
            "name": "com.github.jponge:lzma-java:1.3",
            "downloads": {
                "artifact": {
                    "path": "com/github/jponge/lzma-java/1.3/lzma-java-1.3.jar",
                    "url": "https://files.minecraftforge.net/maven/com/github/jponge/lzma-java/1.3/lzma-java-1.3.jar",
                    "sha1": "a25db9d4d385ccda4825ae1b47a7a61d86e595af",
                    "size": 51041
                }
            }
        },
        {
            "name": "com.google.code.findbugs:jsr305:3.0.2",
            "downloads": {
                "artifact": {
                    "path": "com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar",
                    "url": "https://files.minecraftforge.net/maven/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar",
                    "sha1": "25ea2e8b0c338a877313bd4672d3fe056ea78f0d",
                    "size": 19936
                }
            }
        },
        {
            "name": "com.google.code.gson:gson:2.8.0",
            "downloads": {
                "artifact": {
                    "path": "com/google/code/gson/gson/2.8.0/gson-2.8.0.jar",
                    "url": "https://libraries.minecraft.net/com/google/code/gson/gson/2.8.0/gson-2.8.0.jar",
                    "sha1": "c4ba5371a29ac9b2ad6129b1d39ea38750043eff",
                    "size": 231952
                }
            }
        },
        {
            "name": "com.google.errorprone:error_prone_annotations:2.1.3",
            "downloads": {
                "artifact": {
                    "path": "com/google/errorprone/error_prone_annotations/2.1.3/error_prone_annotations-2.1.3.jar",
                    "url": "https://files.minecraftforge.net/maven/com/google/errorprone/error_prone_annotations/2.1.3/error_prone_annotations-2.1.3.jar",
                    "sha1": "39b109f2cd352b2d71b52a3b5a1a9850e1dc304b",
                    "size": 13704
                }
            }
        },
        {
            "name": "com.google.guava:guava:20.0",
            "downloads": {
                "artifact": {
                    "path": "com/google/guava/guava/20.0/guava-20.0.jar",
                    "url": "https://files.minecraftforge.net/maven/com/google/guava/guava/20.0/guava-20.0.jar",
                    "sha1": "89507701249388e1ed5ddcf8c41f4ce1be7831ef",
                    "size": 2442625
                }
            }
        },
        {
            "name": "com.google.guava:guava:25.1-jre",
            "downloads": {
                "artifact": {
                    "path": "com/google/guava/guava/25.1-jre/guava-25.1-jre.jar",
                    "url": "https://files.minecraftforge.net/maven/com/google/guava/guava/25.1-jre/guava-25.1-jre.jar",
                    "sha1": "6c57e4b22b44e89e548b5c9f70f0c45fe10fb0b4",
                    "size": 2734339
                }
            }
        },
        {
            "name": "com.google.j2objc:j2objc-annotations:1.1",
            "downloads": {
                "artifact": {
                    "path": "com/google/j2objc/j2objc-annotations/1.1/j2objc-annotations-1.1.jar",
                    "url": "https://files.minecraftforge.net/maven/com/google/j2objc/j2objc-annotations/1.1/j2objc-annotations-1.1.jar",
                    "sha1": "ed28ded51a8b1c6b112568def5f4b455e6809019",
                    "size": 8782
                }
            }
        },
        {
            "name": "com.nothome:javaxdelta:2.0.1",
            "downloads": {
                "artifact": {
                    "path": "com/nothome/javaxdelta/2.0.1/javaxdelta-2.0.1.jar",
                    "url": "https://files.minecraftforge.net/maven/com/nothome/javaxdelta/2.0.1/javaxdelta-2.0.1.jar",
                    "sha1": "d4f5d077f1b15c6f67595015d69572303ad3f66d",
                    "size": 38995
                }
            }
        },
        {
            "name": "commons-io:commons-io:2.4",
            "downloads": {
                "artifact": {
                    "path": "commons-io/commons-io/2.4/commons-io-2.4.jar",
                    "url": "https://libraries.minecraft.net/commons-io/commons-io/2.4/commons-io-2.4.jar",
                    "sha1": "b1b6ea3b7e4aa4f492509a4952029cd8e48019ad",
                    "size": 185140
                }
            }
        },
        {
            "name": "de.oceanlabs.mcp:mcp_config:1.13.2-20190213.203750@zip",
            "downloads": {
                "artifact": {
                    "path": "de/oceanlabs/mcp/mcp_config/1.13.2-20190213.203750/mcp_config-1.13.2-20190213.203750.zip",
                    "url": "https://files.minecraftforge.net/maven/de/oceanlabs/mcp/mcp_config/1.13.2-20190213.203750/mcp_config-1.13.2-20190213.203750.zip",
                    "sha1": "b2e3d59e19521e1040e11d97db22d41bb9065a38",
                    "size": 466835
                }
            }
        },
        {
            "name": "net.md-5:SpecialSource:1.8.3",
            "downloads": {
                "artifact": {
                    "path": "net/md-5/SpecialSource/1.8.3/SpecialSource-1.8.3.jar",
                    "url": "https://files.minecraftforge.net/maven/net/md-5/SpecialSource/1.8.3/SpecialSource-1.8.3.jar",
                    "sha1": "9e8fa9e25bfcd93a3a52e1570d8ebf29d4a10799",
                    "size": 78574
                }
            }
        },
        {
            "name": "net.minecraftforge:binarypatcher:1.0.5",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/binarypatcher/1.0.5/binarypatcher-1.0.5.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/binarypatcher/1.0.5/binarypatcher-1.0.5.jar",
                    "sha1": "0743c20fbbc17501da7c700e3241fbbd887d7a20",
                    "size": 16255
                }
            }
        },
        {
            "name": "net.minecraftforge:forge:1.13.2-25.0.191:universal",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/forge/1.13.2-25.0.191/forge-1.13.2-25.0.191-universal.jar",
                    "url": "",
                    "sha1": "407301226b6c434c83912314222fd5168d1cddb8",
                    "size": 2340643
                }
            }
        },
        {
            "name": "net.minecraftforge:installertools:1.0.3",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/installertools/1.0.3/installertools-1.0.3.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/installertools/1.0.3/installertools-1.0.3.jar",
                    "sha1": "594cf5f164ba392cec0dfa9cabd6d7f47a3ca32a",
                    "size": 12115
                }
            }
        },
        {
            "name": "net.minecraftforge:jarsplitter:1.1.0",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/jarsplitter/1.1.0/jarsplitter-1.1.0.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/jarsplitter/1.1.0/jarsplitter-1.1.0.jar",
                    "sha1": "2d26b5fc5b236ba4264dde3d6c3ffcadf03b2020",
                    "size": 7887
                }
            }
        },
        {
            "name": "net.sf.jopt-simple:jopt-simple:4.9",
            "downloads": {
                "artifact": {
                    "path": "net/sf/jopt-simple/jopt-simple/4.9/jopt-simple-4.9.jar",
                    "url": "https://files.minecraftforge.net/maven/net/sf/jopt-simple/jopt-simple/4.9/jopt-simple-4.9.jar",
                    "sha1": "ee9e9eaa0a35360dcfeac129ff4923215fd65904",
                    "size": 66469
                }
            }
        },
        {
            "name": "net.sf.jopt-simple:jopt-simple:5.0.4",
            "downloads": {
                "artifact": {
                    "path": "net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar",
                    "url": "https://files.minecraftforge.net/maven/net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar",
                    "sha1": "4fdac2fbe92dfad86aa6e9301736f6b4342a3f5c",
                    "size": 78146
                }
            }
        },
        {
            "name": "net.sf.opencsv:opencsv:2.3",
            "downloads": {
                "artifact": {
                    "path": "net/sf/opencsv/opencsv/2.3/opencsv-2.3.jar",
                    "url": "https://files.minecraftforge.net/maven/net/sf/opencsv/opencsv/2.3/opencsv-2.3.jar",
                    "sha1": "c23708cdb9e80a144db433e23344a788a1fd6599",
                    "size": 19827
                }
            }
        },
        {
            "name": "org.checkerframework:checker-qual:2.0.0",
            "downloads": {
                "artifact": {
                    "path": "org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.jar",
                    "url": "https://files.minecraftforge.net/maven/org/checkerframework/checker-qual/2.0.0/checker-qual-2.0.0.jar",
                    "sha1": "518929596ee3249127502a8573b2e008e2d51ed3",
                    "size": 343222
                }
            }
        },
        {
            "name": "org.codehaus.mojo:animal-sniffer-annotations:1.14",
            "downloads": {
                "artifact": {
                    "path": "org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.jar",
                    "url": "https://files.minecraftforge.net/maven/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.jar",
                    "sha1": "775b7e22fb10026eed3f86e8dc556dfafe35f2d5",
                    "size": 3482
                }
            }
        },
        {
            "name": "org.ow2.asm:asm-analysis:6.1.1",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm-analysis/6.1.1/asm-analysis-6.1.1.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm-analysis/6.1.1/asm-analysis-6.1.1.jar",
                    "sha1": "376d85732456ec4d2c4bc020ee152061c61f00b5",
                    "size": 31979
                }
            }
        },
        {
            "name": "org.ow2.asm:asm-commons:6.1.1",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm-commons/6.1.1/asm-commons-6.1.1.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm-commons/6.1.1/asm-commons-6.1.1.jar",
                    "sha1": "70a7deee986e66c015cea1c9978f9654cc667020",
                    "size": 77745
                }
            }
        },
        {
            "name": "org.ow2.asm:asm-tree:6.1.1",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm-tree/6.1.1/asm-tree-6.1.1.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm-tree/6.1.1/asm-tree-6.1.1.jar",
                    "sha1": "19fdc358e7bd3986b9867de508ce6cefa693dda6",
                    "size": 50050
                }
            }
        },
        {
            "name": "org.ow2.asm:asm:6.1.1",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm/6.1.1/asm-6.1.1.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm/6.1.1/asm-6.1.1.jar",
                    "sha1": "264754515362d92acd39e8d40395f6b8dee7bc08",
                    "size": 108252
                }
            }
        },
        {
            "name": "trove:trove:1.0.2",
            "downloads": {
                "artifact": {
                    "path": "trove/trove/1.0.2/trove-1.0.2.jar",
                    "url": "https://files.minecraftforge.net/maven/trove/trove/1.0.2/trove-1.0.2.jar",
                    "sha1": "112b40581ab9a4d3de4636d49985a7f686d181c8",
                    "size": 361155
                }
            }
        }
    ]
}
```

Example `version.json`:

```json
{
    "_comment_": [
        "Please do not automate the download and installation of Forge.",
        "Our efforts are supported by ads from the download page.",
        "If you MUST automate this, please consider supporting the project through https://www.patreon.com/LexManos/"
    ],
    "id": "1.13.2-forge-25.0.191",
    "time": "2019-04-25T23:56:20+00:00",
    "releaseTime": "2019-04-25T23:56:20+00:00",
    "type": "release",
    "mainClass": "cpw.mods.modlauncher.Launcher",
    "inheritsFrom": "1.13.2",
    "logging": {
        
    },
    "arguments": {
        "game": [
            "--launchTarget",
            "fmlclient",
            "--fml.forgeVersion",
            "25.0.191",
            "--fml.mcVersion",
            "1.13.2",
            "--fml.forgeGroup",
            "net.minecraftforge",
            "--fml.mcpVersion",
            "20190213.203750"
        ]
    },
    "libraries": [
        {
            "name": "net.minecraftforge:forge:1.13.2-25.0.191",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/forge/1.13.2-25.0.191/forge-1.13.2-25.0.191.jar",
                    "url": "",
                    "sha1": "66563952b27e32cda3bc5ac5db1cfd54f87771df",
                    "size": 153985
                }
            }
        },
        {
            "name": "com.paulscode:soundsystem:201809301515",
            "downloads": {
                "artifact": {
                    "path": "com/paulscode/soundsystem/201809301515/soundsystem-201809301515.jar",
                    "url": "https://files.minecraftforge.net/maven/com/paulscode/soundsystem/201809301515/soundsystem-201809301515.jar",
                    "sha1": "383fd2cc26b681583d9e75ab4c20263df0387093",
                    "size": 66607
                }
            }
        },
        {
            "name": "org.ow2.asm:asm:6.2",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm/6.2/asm-6.2.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm/6.2/asm-6.2.jar",
                    "sha1": "1b6c4ff09ce03f3052429139c2a68e295cae6604",
                    "size": 111214
                }
            }
        },
        {
            "name": "org.ow2.asm:asm-commons:6.2",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm-commons/6.2/asm-commons-6.2.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm-commons/6.2/asm-commons-6.2.jar",
                    "sha1": "f0df1c69e34a0463679d7c8db36ddb4312836e76",
                    "size": 78919
                }
            }
        },
        {
            "name": "org.ow2.asm:asm-tree:6.2",
            "downloads": {
                "artifact": {
                    "path": "org/ow2/asm/asm-tree/6.2/asm-tree-6.2.jar",
                    "url": "https://files.minecraftforge.net/maven/org/ow2/asm/asm-tree/6.2/asm-tree-6.2.jar",
                    "sha1": "61570e046111559f38d4e0e580c005f75988c0a6",
                    "size": 50370
                }
            }
        },
        {
            "name": "cpw.mods:modlauncher:1.0.0",
            "downloads": {
                "artifact": {
                    "path": "cpw/mods/modlauncher/1.0.0/modlauncher-1.0.0.jar",
                    "url": "https://files.minecraftforge.net/maven/cpw/mods/modlauncher/1.0.0/modlauncher-1.0.0.jar",
                    "sha1": "1bda9e20394adf7c65e80f54c07fa233025275a1",
                    "size": 87486
                }
            }
        },
        {
            "name": "net.minecraftforge:accesstransformers:0.16.0-shadowed",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/accesstransformers/0.16.0-shadowed/accesstransformers-0.16.0-shadowed.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/accesstransformers/0.16.0/accesstransformers-0.16.0-shadowed.jar",
                    "sha1": "a87a163b87fd3ce583038f56d4e8bb9c97ca5228",
                    "size": 444428
                }
            }
        },
        {
            "name": "net.minecraftforge:eventbus:0.8.1-service",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/eventbus/0.8.1-service/eventbus-0.8.1-service.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/eventbus/0.8.1/eventbus-0.8.1-service.jar",
                    "sha1": "d83d68c44a0db06f780aa514767537b7f2788eae",
                    "size": 37341
                }
            }
        },
        {
            "name": "net.minecraftforge:forgespi:0.11.0",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/forgespi/0.11.0/forgespi-0.11.0.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/forgespi/0.11.0/forgespi-0.11.0.jar",
                    "sha1": "304517500e4422e709ef1a1d48dd5a77f1634795",
                    "size": 16076
                }
            }
        },
        {
            "name": "net.minecraftforge:coremods:0.4.2",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/coremods/0.4.2/coremods-0.4.2.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/coremods/0.4.2/coremods-0.4.2.jar",
                    "sha1": "8a56273d9f638f67d0e2fb5aafa0972557f4c2f5",
                    "size": 11264
                }
            }
        },
        {
            "name": "net.minecraftforge:unsafe:0.2.0",
            "downloads": {
                "artifact": {
                    "path": "net/minecraftforge/unsafe/0.2.0/unsafe-0.2.0.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecraftforge/unsafe/0.2.0/unsafe-0.2.0.jar",
                    "sha1": "54d7a0a5e8fdb71b973025caa46f341ae5904f39",
                    "size": 2834
                }
            }
        },
        {
            "name": "com.electronwill.night-config:core:3.6.0",
            "downloads": {
                "artifact": {
                    "path": "com/electronwill/night-config/core/3.6.0/core-3.6.0.jar",
                    "url": "https://files.minecraftforge.net/maven/com/electronwill/night-config/core/3.6.0/core-3.6.0.jar",
                    "sha1": "0412f2edf0ef6cc9178ab38751af4e507f413bef",
                    "size": 199763
                }
            }
        },
        {
            "name": "com.electronwill.night-config:toml:3.6.0",
            "downloads": {
                "artifact": {
                    "path": "com/electronwill/night-config/toml/3.6.0/toml-3.6.0.jar",
                    "url": "https://files.minecraftforge.net/maven/com/electronwill/night-config/toml/3.6.0/toml-3.6.0.jar",
                    "sha1": "f0c2c48748bdfc8d1f3798ec478f003bccc7e0b1",
                    "size": 31256
                }
            }
        },
        {
            "name": "org.jline:jline:3.9.0",
            "downloads": {
                "artifact": {
                    "path": "org/jline/jline/3.9.0/jline-3.9.0.jar",
                    "url": "https://files.minecraftforge.net/maven/org/jline/jline/3.9.0/jline-3.9.0.jar",
                    "sha1": "da6eb8ebdd131ec41f7e42e7e77b257868279698",
                    "size": 707273
                }
            }
        },
        {
            "name": "org.apache.maven:maven-artifact:3.6.0",
            "downloads": {
                "artifact": {
                    "path": "org/apache/maven/maven-artifact/3.6.0/maven-artifact-3.6.0.jar",
                    "url": "https://files.minecraftforge.net/maven/org/apache/maven/maven-artifact/3.6.0/maven-artifact-3.6.0.jar",
                    "sha1": "d4c0da647de59c9ccc304a112fe1f1474d49e8eb",
                    "size": 55051
                }
            }
        },
        {
            "name": "net.jodah:typetools:0.6.0",
            "downloads": {
                "artifact": {
                    "path": "net/jodah/typetools/0.6.0/typetools-0.6.0.jar",
                    "url": "https://files.minecraftforge.net/maven/net/jodah/typetools/0.6.0/typetools-0.6.0.jar",
                    "sha1": "a1552ecd6f5b9444585a7c4b05f6312cf7d32fd3",
                    "size": 14734
                }
            }
        },
        {
            "name": "java3d:vecmath:1.5.2",
            "downloads": {
                "artifact": {
                    "path": "java3d/vecmath/1.5.2/vecmath-1.5.2.jar",
                    "url": "https://libraries.minecraft.net/java3d/vecmath/1.5.2/vecmath-1.5.2.jar",
                    "sha1": "79846ba34cbd89e2422d74d53752f993dcc2ccaf",
                    "size": 318956
                }
            }
        },
        {
            "name": "org.apache.logging.log4j:log4j-api:2.11.1",
            "downloads": {
                "artifact": {
                    "path": "org/apache/logging/log4j/log4j-api/2.11.1/log4j-api-2.11.1.jar",
                    "url": "https://files.minecraftforge.net/maven/org/apache/logging/log4j/log4j-api/2.11.1/log4j-api-2.11.1.jar",
                    "sha1": "268f0fe4df3eefe052b57c87ec48517d64fb2a10",
                    "size": 264060
                }
            }
        },
        {
            "name": "org.apache.logging.log4j:log4j-core:2.11.1",
            "downloads": {
                "artifact": {
                    "path": "org/apache/logging/log4j/log4j-core/2.11.1/log4j-core-2.11.1.jar",
                    "url": "https://files.minecraftforge.net/maven/org/apache/logging/log4j/log4j-core/2.11.1/log4j-core-2.11.1.jar",
                    "sha1": "592a48674c926b01a9a747c7831bcd82a9e6d6e4",
                    "size": 1607947
                }
            }
        },
        {
            "name": "net.minecrell:terminalconsoleappender:1.1.1",
            "downloads": {
                "artifact": {
                    "path": "net/minecrell/terminalconsoleappender/1.1.1/terminalconsoleappender-1.1.1.jar",
                    "url": "https://files.minecraftforge.net/maven/net/minecrell/terminalconsoleappender/1.1.1/terminalconsoleappender-1.1.1.jar",
                    "sha1": "d7e48a3c5f778bb8b41d178a52b4411dff418a0c",
                    "size": 15240
                }
            }
        },
        {
            "name": "net.sf.jopt-simple:jopt-simple:5.0.4",
            "downloads": {
                "artifact": {
                    "path": "net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar",
                    "url": "https://files.minecraftforge.net/maven/net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar",
                    "sha1": "4fdac2fbe92dfad86aa6e9301736f6b4342a3f5c",
                    "size": 78146
                }
            }
        }
    ]
}
```

**Comment by RyanTheAllmighty** [comment]
Alright so I've pushed up code that gets this working with latest forge.

I need to massive clean up, cache some things and check local hashes as well as make sure it didn't break anything else.

Also need to check old forge versions and see if they can also be automated the same way, just without the processors

**Comment by RyanTheAllmighty** [comment]
Work is done and tested working for Minecraft versions 1.10 and above with Forge.

Still to do is:

- Server installs
- Updating from an old instance pre this new release, to a version using the new `<loader />` syntax
- Forge XZ downloads don't seem to be caching correctly and redownload every time, so that needs to be fixed
- Add ability for using `<loader type="forge" recommended />` and  `<loader type="forge" latest />`. Packs will most likely want to set a specific version or Forge, but might be good for quick testing.
- Add ability for selecting any version of Forge (or none) for the Vanilla pack (or any I guess for that matter) by using something like `<loader type="forge" select />`

**Comment by RyanTheAllmighty** [comment]
Supporting Fabric loader:

- Download https://fabricmc.net/download/vanilla?yarn=${yarn}&loader=${loader}
- Extract zip and extract `fabric-loader-${yarn}-${loader}.json`
- Use that JSON as a normal Minecraft profile json, it extends normal Minecraft version so it's libraries and arguments are additive and mainclass is overwrite.

Example xml usage:

`<loader type="fabric" yarn="1.14+build.7" loader="0.4.6+build.141" />`

**Comment by RyanTheAllmighty** [comment]
Alright Fabric support is in as well as using `latest` and `recommended` for forge.

Work is done and tested working for Minecraft versions 1.10 and above with Forge.

Still to do is:

- Server installs
- Updating from an old instance pre this new release, to a version using the new `<loader />` syntax
- Forge XZ downloads don't seem to be caching correctly and redownload every time, so that needs to be fixed

**Comment by RyanTheAllmighty** [comment]
Alright all the issues seem to be fixed now.

I've tested extensively including:

- Forge 1.13.2 client and server installs
- Forge < 1.13.2 client and server installs
- Installing a pack not using the `<loader />` xml
- Installing a pack using the `<loader />` xml
- Installing a pack not using the `<loader />` xml from an old launcher, then updating to a version of the pack using the `<loader />` syntax in the new launcher
- Fabric client and server installs

No issues are known. I'll continue testing tonight and get some more pack devs to help on discord, and then release tomorrow if all is well.

**Comment by RyanTheAllmighty** [comment]
Final syntax is:

For Forge:
`<loader type="forge" version="version_number" />`
`<loader type="forge" latest />`
`<loader type="forge" recommended />`

For Fabric:
`<loader type="fabric" yarn="version_number" loader="version_number" />`
`<loader type="fabric" latest="true" />`

**Comment by 7osCraft** [comment]
What about Lite Loader and the ability to use 2 loaders like forge + liteloader?

**Comment by Y0ungSandwich** [comment]
> What about Lite Loader and the ability to use 2 loaders like forge + liteloader?

Already possible.

**Comment by 7osCraft** [comment]
@Y0ungSandwich But I mean with this new loader system/method?

**Comment by RyanTheAllmighty** [comment]
Ah. I meant to look.

I'll get that in tonight for tomorrow's release

**Comment by RyanTheAllmighty** [comment]
Multiple loaders may be a future thing, but for simplicity and to cover the 99% use case, a single loader is fine.

Do forge and liteloader even work together?

**Comment by 7osCraft** [comment]
Yes, they work together.
The only use case I'd think of is using voxel map but there is alot of minimaps  to use.

**Comment by Y0ungSandwich** [comment]
Yeah we were using it for a bit back in the day for voxel map.
https://wiki.atlauncher.com/how_do_i:add_liteloader?s[]=liteloader

**Comment by jamierocks** [comment]
I don't believe Liteloader is / has been maintained for a while anyhow, might be fitting for the legacy system.

Definitely something to consider having though, multiple loaders that is.

**Comment by RyanTheAllmighty** [comment]
Released with 3.2.5.0 just now
